[Sema] Improve error for unqualified use of static variable in method

aeu opened the pull request:
### Resolves 
https://github.com/swiftlang/swift/issues/48759

### Summary
This PR improves the diagnostic emitted when a static member is referenced through an instance.
The previous diagnostic was misleading because it stated that the static member “cannot be used on an instance,” without clarifying the correct usage.
This change updates the diagnostic to explicitly state that static members must be referenced on the type.

### Changes made

1. Added a new diagnostic to include/swift/AST/DiagnosticsSema.def
2. Updated lib/Sema/CSDiagnostics.cpp to extract the relevant tokens and emit the updated message
3. Added test coverage in test/Sema/static_member_on_instance.swift


### Original Example (from Issue)
````
struct HasStatic {
    func foo() {
        print(cvar)
    }
    static let cvar = 123
}
````

### Previous Output
````
$ swiftc 48759.swift 
48759.swift:3:15: error: static member 'cvar' cannot be used on instance of type 'HasStatic'
1 | struct HasStatic {
2 |     func foo() {
3 |         print(cvar)
  |               `- error: static member 'cvar' cannot be used on instance of type 'HasStatic'
4 |     }
5 |     static let cvar = 123
````

### Updated Output
````
$ swiftc 48759.swift 
48759.swift:3:15: error: static member 'cvar' can only be used on the type 'HasStatic', not on the instance self
1 | struct HasStatic {
2 |     func foo() {
3 |         print(cvar)
  |               `- error: static member 'cvar' can only be used on the type 'HasStatic', not on the instance self
4 |     }
5 |     static let cvar = 123
````

aeu committed:
[Sema] Improve error for unqualified use of static variable in method

xedin reviewed (changes requested):
Thank you for the contribution! I left a comment inline regarding the code itself and I think the new wording might not actually make things clearer.

aeu commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -5068,8 +5068,33 @@ bool AllowTypeOrInstanceMemberFailure::diagnoseAsError() {
+          auto SR = semantic->getSourceRange();
+          if (SR.isValid()) {
+            auto CSR = Lexer::getCharSourceRangeFromSourceRange(SM,SR);
+            instanceName = SM.extractText(CSR);
```
> ... but regardless of that this is not the way to get the instance reference, it should be recorded as part of recording a fix.

Hi @xedin  - If I am reading your suggestion correctly, I should create a fix and record it with the base during constraint solving and then have the emitter read it rather than pulling it from the source text?

aeu committed:
48759 - incorporate PR comments

aeu commented:
Commit https://github.com/swiftlang/swift/pull/85706/commits/84d0f08d0356ff3d6d5576875cd8c4783628712e addresses comments by @xedin regarding how to get the instance reference.   

The instance expression is captured when the fix is created, and the diagnostic emission uses it.  This is now following the patterns used by AllowInvalidMemberRef / MemberReferenceFailure and TreatRValueAsLValue.

Additionally the text expectations were updated to the new diagnostic.

aeu committed:
48759 - make copy in tests consistent with new diagnostic

aeu committed:
48759 - capture clang changes
